databuilder/query_engines/mssql_dialect.py

```diff
@@ -69,12 +69,29 @@ class MSSQLDateTime(_MSSQLDateTimeBase, sqlalchemy.types.TypeDecorator):
     format_str = "%Y-%m-%dT%H:%M:%S"
 
 
+# MS-SQL can interpret values that we intend to be floats as Decimals, which
+# can result in results with unexpectedly truncated precision.
+# https://github.com/opensafely-core/databuilder/issues/1065
+# https://learn.microsoft.com/en-us/sql/t-sql/data-types/precision-scale-and-length-transact-sql
+class MSSQLFloat(sqlalchemy.types.TypeDecorator):
+    impl = sqlalchemy.types.Float
+    cache_ok = True
+
+    def bind_expression(self, bindvalue):
+        # Wrap any bound parameters in an explicit CAST to their intended type.
+        # This ensures that any float-like numbers are always cast to
+        # floats, otherwise MSSQL will sometimes make a best-guess and return a
+        # Decimal type, which behaves differently to other engine.
+        return cast(bindvalue, type_=self)
```

Comment:
I didn't initially think of doing this via CAST, although this ought to work fine.

For what it's worth, MSSQL has a way of expressing float literals using scientific notation:
https://learn.microsoft.com/en-us/sql/t-sql/data-types/constants-transact-sql?view=sql-server-2017#float-and-real-constants

So I think we could do this by adding a `process_bind_param()` method which formats the Python float as a string and then sticks `E0` on the end (if the string doesn't contain `E` already).

I think this would result in slightly cleaner SQL, but there's probably not much practical difference.

Reply:
@evansd I've tried this, and I couldn't get it to work.  If I format it as a string (in `process_bind_param()`), it complains about types:
```
sqlalchemy.exc.OperationalError: (pymssql._pymssql.OperationalError) (8117, b'Operand data type nvarchar is invalid for sum operator.DB-Lib error message 20018, severity 16:\nGeneral SQL Server error: Check messages from the SQL Server\n')
[SQL: SELECT sum(%(sum_1)s) + %(sum_2)s AS anon_1]
[parameters: {'sum_1': '0.5E0', 'sum_2': '0.25E0'}]
```
If I return it as the literal `0.5E0` (couldn't actually figure out how to do that, but I made it return the constants as a test), it still ends up with the truncated value (I guess because it means the same as `0.5`, so it hasn't actually been converted to anything different).

Reply:
Ah, yeah I see. I guess there's no way round that then given the limitations of `pymssql`. Thanks for investigating.